Disallow Encode(payload) with AddClaim(s)

hartmark opened the pull request:
Fixes for https://github.com/jwt-dotnet/jwt/issues/456.

hartmark commented:
Fix for #456

abatishchev commented on src/JWT/Builder/JwtBuilder.cs:
```diff
@@ -468,5 +474,42 @@ private bool CanDecodeHeader()
+#endif
+                    default:
+                    {
+                        throw new NotSupportedException($"{jsonSerializer.GetType()} is not supported");
```
I don't think we should throw an exception here. It defeat the purpose of having an interface, we don't know if the customers have their own implementation, the idea is that they might.

abatishchev replied:
Can the `default` case be just omitted so the method would return `prop.Name`?

hartmark replied:
good idea

abatishchev commented on src/JWT/Builder/JwtBuilder.cs:
```diff
@@ -276,7 +282,7 @@ public string Encode(Type payloadType, object payload)
             EnsureCanEncode();
 
             var dic = payloadType.GetProperties(BindingFlags.Instance | BindingFlags.Public)
```
Overall, I think this is not the rich approach, the library shouldn't try to reinvent the wheel, must somehow let the serializer itself to deal with the type of the payload.

abatishchev replied:
@hartmark what you think about this variant of the method?

```c#
  public string Encode(Type payloadType, object payload)
  {
      if (payloadType is null)
          throw new ArgumentNullException(nameof(payloadType));
      if (payload is null)
          throw new ArgumentNullException(nameof(payload));
      
      EnsureCanEncode();

      return _encoder.Encode(_jwt.Header, payload, _secrets?[0]);
  }
```

which passes the newly added tests too!

hartmark replied:
The tests that just do .Encode without .AddClaim works

For example Encode_Should_Return_Token_With_Custom_Extra_Headers_Full_Payload_And_Claims fails as it adds the claim "ExtraClaim" first and that gets overwritten by .Encode method.

abatishchev reviewed (commented):
Ah, it ran into a merge conflict now :(

abatishchev commented:
Do I understand the origin of the issue correct is that it occurs only in the builder because its Encode() method "messed up" with the payload so a complex model would work?

If that's the case, what I think we should do: remove the code and move it to a custom serializer, which in turn just a decorator on the other one (maybe two decorators, one for the each underlying technology).

hartmark commented:
I'll check the conflict when I'm at a computer. Looks like something minor.

Yes, the issue is that Encode() with a parameter serializes the _jwt.Payload property.

While Encode() with a parameter serializes only what we pass to it, ignoring the _jwt.Payload if we don't either add what we pass to it or adds the claims manually.

hartmark commented:
I have reworked the Encode when also having added Claims so they are both serialized the same way as everything else.

I have for now just added another Encode in JwtEncoder that just accepts two payloads. And in that case just serializes them and uses the merge functionallity in JSON.net, for System.Text.Json it was a bit more tricky but after finding out you can use the JsonNode classes it was easier.

I think this is a more clean solution than iterating the payload with reflection and add all properties in the _jwt.Payload dictionary in JwtBuilder.

If this is an acceptable solution we could rework the JwtEncoder.Encode method to be less copy.paste

All tests pass also! :)

PS. For further reading about JsonNode and Merge
https://github.com/dotnet/runtime/issues/31433
https://makolyte.com/csharp-how-to-use-jsonnode-to-read-write-and-modify-json/

abatishchev commented on src/JWT/IJsonSerializer.cs:
```diff
@@ -21,6 +21,14 @@ public interface IJsonSerializer
+        /// <param name="obj1">The first object to serialize</param>
+        /// <param name="obj2">The second object to serialize, note that any existing property in obj1 will be
+        /// <returns>JSON string</returns>
+        string MergeObjects(object obj1, object obj2);
```
I feel it's too hacky, in other words - leaky abstraction. Is there a way to support a complex model without adding this method?

hartmark replied:
Fair enough, the solution before this commit also works so if that's a more acceptable solution we can revert back to that.

abatishchev replied:
Yes, please. Let ship a beta version and it works for everyone - then let's keep it simple.

However i cannot stress enough my appreciation for your contributions and trying to make the library better over and over again! hope you're ok with my ask to revert.

hartmark replied:
Alright, no worries. It was an interesting solution but perhaps not the smallest code change 😀

I'll try get this sorted when I get the time, hopefully this evening.

hartmark replied:
I have reverted that commit and rebased against latest on main now

hartmark commented on src/JWT/Builder/JwtBuilder.cs:
```diff
@@ -272,18 +278,16 @@ public string Encode(Type payloadType, object payload)
-
-            foreach (var pair in dic)
+            // Preserve old behaviour checking the type, is it needed?
+            if (payloadType != payload.GetType())
```
I only added this because once test checked this. The previous implementation needed to know the type to be able to deconstruct the object with reflection.

The new implementation just uses generic serialisation so it's no longer needed to add the type.

abatishchev replied:
I'd keep the library lightweight and relying on the underlying serialier to fail

abatishchev commented on src/JWT/Builder/JwtBuilder.cs:
```diff
@@ -276,7 +282,7 @@ public string Encode(Type payloadType, object payload)
             EnsureCanEncode();
 
             var dic = payloadType.GetProperties(BindingFlags.Instance | BindingFlags.Public)
```
I still feel this is hacky. Is it possible to avoid doing it altogether?

hartmark replied:
If we disallow using both .AddClaim()/.AddClaims() and ,Encode(payload) we should then be able to just send the payload directly to the serializer.

hartmark replied:
Yes, that sounds like a reasonable approach

hartmark replied:
I have added code to disallow mixing Encode(payload) with AddClaim/AddClaims

abatishchev commented:
Awesome, thank you!! Pushed some minor formatting changes. And dropped the beta prefix, I think it'll work as-is.

hartmark commented:
Cool, that was fast. I think we ended up with a cleaner solution with just sacrificing a minor nice-to-have functionallity
